core/src/worker/workflow/machines/activity_state_machine.rs

```diff
@@ -394,19 +401,24 @@ impl ScheduleCommandCreated {
         dat: SharedState,
         sched_dat: ActTaskScheduledData,
     ) -> ActivityMachineTransition<ScheduledEventRecorded> {
-        if sched_dat.act_id != dat.attrs.activity_id {
-            return TransitionResult::Err(WFMachinesError::Nondeterminism(format!(
-                "Activity id of scheduled event '{}' does not \
+        if dat.internal_flags.borrow_mut().try_use(
+            CoreInternalFlags::IdAndTypeDeterminismChecks,
+            sched_dat.last_task_in_history,
```

Comment:
Scenario I just thought up:

* I have a workflow started a while back that has started an activity and is waiting on it (so schedule activity is in the last task)
* I've since updated my workflow code to rename the activity type
* I issue a query against the this long-since running workflow and it works against my older SDK because this check is not in place
* I upgrade my SDK
* I issue a query again and it breaks now because the activity is technically the last task in history

Any concerns with this scenario?

Reply:
Just noting that having the check here means that the flag will essentially be added on every workflows calling an activity from this point on. Alternatively, it would be possible to do this check only if one of the check below fails, thus adding the flag only on workflows that actually have a problem.

Nothing bad here, just mentioning because I think it somehow contradicts something else you wrote elsewhere, but can't find back now, so feel free to ignore.

Reply:
@cretz I'm OK with that scenario. The workflow hasn't yet ever "verified" the activity, because it's never processed that latest task. In fact, I'm not even totally sure your scenario can happen, because the first query would I believe cause the old worker to process and complete the task containing the schedule event, and then then new worker wouldn't even run the check.